i6_models/primitives/specaugment.py

```diff
@@ -0,0 +1,123 @@
+import torch
+
+
+def _mask(tensor: torch.Tensor, batch_axis: int, axis: int, pos: torch.Tensor, max_len: int) -> torch.Tensor:
+    """
+    :param tensor: e.g. [B, ..., A, ...] but arbitrary axis order
+    :param batch_axis: index of the batch axis
+    :param axis: which axis A to mask
+    :param pos: at which positions along axis to start the mask (size [B])
+    :param max_len: mask length drawn uniformly from [0, max_len]
+    """
+    batch_dim = tensor.shape[batch_axis]
+    dim = tensor.shape[axis]
+    amount = torch.randint(low=1, high=max_len + 1, size=(batch_dim,), dtype=torch.int32, device=tensor.device)
+    pos2 = torch.min(pos + amount, torch.tensor([dim] * batch_dim, device=tensor.device))
+    idxs = torch.arange(0, dim, device=tensor.device).unsqueeze(0)  # [1,dim]
+    pos_bc = pos.unsqueeze(1)  # [B,1]
+    pos2_bc = pos2.unsqueeze(1)  # [B,1]
+    cond = torch.logical_and(torch.greater_equal(idxs, pos_bc), torch.less(idxs, pos2_bc))  # [B,dim]
+    if batch_axis > axis:
+        cond = cond.transpose(0, 1)  # [dim,B]
+    cond = torch.reshape(
+        cond, shape=[tensor.shape[i] if i in (batch_axis, axis) else 1 for i in range(len(tensor.shape))]
+    )
+    tensor = torch.where(cond, 0.0, tensor)
+    return tensor
+
+
+def _random_mask(tensor: torch.Tensor, batch_axis: int, axis: int, min_num: int, max_num: int, max_len: int):
+    """
+    Mask tensor along axis using N in [min_num, max_num] masks of length [0, max_len]
+
+    :param tensor: e.g. [B, ..., A, ...] but arbitrary axis order
+    :param batch_axis: index of the batch axis
+    :param axis: which axis to mask
+    :param min_num: minimum number of masks
+    :param max_num: maximum number of masks
+    :param max_amount: mask length drawn uniformly from [0, max_amount]
+    """
+
+    batch_dim = tensor.shape[batch_axis]
+    num_masks = torch.randint(min_num, max_num, size=(batch_dim,), device=tensor.device)  # [B]
+
+    z = torch.rand((batch_dim, tensor.shape[axis]), device=tensor.device)  # [B,dim]
+    _, indices = torch.topk(z, num_masks.max().item(), dim=1)
+
+    for i in range(num_masks.max().item()):
+        tensor = _mask(tensor, batch_axis, axis, indices[:, i], max_len)
```

Comment:
I just notice, this is not like in the original implementation, where we do:
```python
_, x = tf.while_loop(
    cond=lambda i, _: tf.less(i, tf.reduce_max(num)),
    body=lambda i, x: (
        i + 1,
        tf.where(             # <
            tf.less(i, num),  # <--- this here is missing
            _mask(x, batch_axis=batch_axis, axis=axis, pos=indices[:, i], max_amount=max_dims, mask_value=mask_value),
            x)),
    loop_vars=(0, x))
```

Reply:
Yes, that makes much more sense. With this condition the actual values in `num_masks` will be used and not just the max. Something like this would be the torch equivalent:
```
    # Make num_masks broadcastable to shape of tensor for torch.where.
    for _ in range(tensor.dim() - 1):
        num_masks = num_masks.unsqueeze(-1)

    for i in range(num_masks.max().item()):
        tensor = torch.where(
            i < num_masks,
            _mask(tensor, batch_axis, axis, indices[:, i], max_len),
            tensor
        )
```

Reply:
Ah, did not read this part yet, now I understand.

Reply:
I'm not sure if broadcasting is correct in the code, but otherwise, yes.

Reply:
this broadcasting is not correct if batch_axis != 0, but otherwise this should be ok.

Reply:
> this broadcasting is not correct if batch_axis != 0, but otherwise this should be ok.

Ah that's true. In the general case
```
    for _ in range(batch_axis, tensor.dim() - 1):
        num_masks = num_masks.unsqueeze(-1)
```
should do the trick.